fix(appium): fix extension autoinstall postinstall script

boneskull opened the pull request:
Resolves #16924

* * *

- This is a little more involved than I would have liked.  Maybe it'd make more sense to just spawn e.g., `appium driver install` child processes?
- Since this script exits with a non-zero code, it can prevent installation of Appium altogether, which seems undesirable.

boneskull committed:
fix(appium): fix extension autoinstall postinstall script

Resolves #16924

Because the `postinstall` script of `appium` needs functions in `@appium/support`, that module must be compiled _before_ its `postinstall` script runs.  Prior to this PR, only the monorepo root would run a `postinstall` script to build the subpackages.  Now, each package has a `prepare` script which runs its own build; this happens prior to any `postinstall`.  This way, the `postinstall` script in `appium` can run successfully.

mykola-mokhnach commented on packages/appium/scripts/postinstall.js:
```diff
@@ -14,20 +17,34 @@ async function main () {
+    [PLUGIN_TYPE, pluginEnv],
+  ];
+
+  const appiumHome = process.env.npm_config_appium_home || (await env.resolveAppiumHome());
```
shouldn't `npm_config_appium_home` be also considered in `resolveAppiumHome` call?

boneskull replied:
I'm not sure what you're asking.  Are you asking if `resolveAppiumHome()` should consider that environment variable in addition to `APPIUM_HOME`?

mykola-mokhnach replied:
yes, exactly

jlipps replied:
i think we settled on just leaning on APPIUM_HOME. i think this code came from a time when we had the `--appium-home` command line flag, and it mirrors that. but we no longer have that flag.

mykola-mokhnach commented on packages/appium/scripts/postinstall.js:
```diff
@@ -36,35 +53,41 @@ async function main () {
+      suppressOutput: true,
+      [type]: ext,
+    },
+    type === 'driver' ? driverConfig : pluginConfig
```
magic string?

boneskull commented:
I've updated the `postinstall.js` script to _never_ exit with a non-zero code; errors will be displayed, but the installation will continue.

jlipps commented on packages/appium/scripts/postinstall.js:
```diff
@@ -10,23 +23,15 @@ async function main() {
   ];
 
+  const appiumHome =
+    process.env.npm_config_appium_home || (await env.resolveAppiumHome());
```
i *think* we can do away with this appium home npm install flag and just rely on APPIUM_HOME or the standard env discovery.

boneskull replied:
OK.  Though ... I see a problem I hadn't thought of before.

This script is only applicable if Appium _isn't_ a direct dependency within the closest `package.json`. If that's the case, then it's expected these drivers/plugins will _also_ be direct dependencies.

I'll address this

jlipps reviewed (approved):
i'm not sure if the right thing is actually to prevent installation of appium? i mean if someone requests drivers as part of the install, usually in a CI environment, wouldn't they want to know if those drivers werent installed before running their tests?

boneskull commented:
@jlipps 

> i'm not sure if the right thing is actually to prevent installation of appium? i mean if someone requests drivers as part of the install, usually in a CI environment, wouldn't they want to know if those drivers werent installed before running their tests?

How about a compromise: fail if `process.env.CI`?

boneskull commented:
I made it nicer.   Here's some screenshots, but there's also _spinners_:

![image](https://user-images.githubusercontent.com/924465/167720439-c76b3150-02db-4971-91ed-7298fb3b9c04.png)

![image](https://user-images.githubusercontent.com/924465/167720475-11dc4453-2f45-4a10-8c22-549ff2d005fb.png)

![image](https://user-images.githubusercontent.com/924465/167720594-bef105d0-cccf-4140-8ffb-bfae463b094e.png)

jlipps commented on packages/appium/scripts/postinstall.js:
```diff
@@ -1,83 +1,162 @@
-      `Could not load extension CLI file; has the project been transpiled? ` + `(${e.message})`
+  if (!driverEnv && !pluginEnv) {
+    spinner.succeed(
+      `No drivers or plugins to automatically install. If desired, set the "npm_config_drivers" or "npm_config_plugins" environment variable(s) to a comma-separated list, and they will be installed automatically with Appium.`
```
NPM sets these env vars. What the user does is add them as flags to npm install, e.g.,

```
npm install -g appium --drivers=fake,xcuitest
```

then NPM magically turns `--drivers` into `process.env.npm_config_drivers`.

all that to say, we should probably refer to the `--drivers` and `--plugins` way of doing things because that is the intended user facing feature.

boneskull replied:
Ah, k.

jlipps reviewed (changes requested):
looks good, one comment

jlipps reviewed (approved):
LGTM when CI passes

boneskull commented:
ugh what have I done

boneskull commented:
OK, so basically: we can't require anything from `@appium/support` here because it won't have been built yet.  The `postinstall` scripts in the subpackages run before the postinstall script in the monorepo.  I think what we can do is add `prepare` scripts to each subpackage which call `npm run build`. These `prepare` scripts only get called before publish and on an `npm install` in a working copy.

boneskull committed:
test(appium): bump exec time of cli e2e tests to avoid flake

boneskull committed:
chore(appium): renamed postinstall.js to autoinstall-extensions.js

...just so we're clear about what it does.  Added some more explanation
in docstrings.

boneskull commented:
Current dependencies on/for this PR:
* 2.0
  * **PR #16902** <a href="https://app.graphite.dev/github/pr/appium/appium/16902" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/appium/appium/16925?utm_source=stack-comment).

boneskull commented on package.json:
```diff
@@ -98,7 +97,8 @@
     "@appium/universal-xml-plugin": "file:packages/universal-xml-plugin",
-    "appium": "file:packages/appium"
+    "appium": "file:packages/appium",
+    "lru-cache": "7.9.0"
```
hmm, I was trying to fix the `lru-cache` bug.  I'll rip that out and try again.

boneskull replied:
yeah, wasn't needed.
